### PR TITLE
(#1397) Create combo box

### LINKED ISSUE #1397: Story: Create the NCI Combo Box
### Requirements

- The nci-combo-box will be created
- The nci-combo-box will have the same features and functionality as the USWDS component with the addition of the following changes:
  - Accessibility text will be provided for the combo in a non-visible `<div>` with `role=status` to indicate the contents of the combo box for screen reader users
  - When the user clicks or tabs into the combo box, the first five options from the associated list will appear in the dropdown
  - When a user types a full matching word/phrase by characters (capitalization does not matter) and presses enter, then the selection will appear in the combo box
  - If a user types a partial matching word and presses enter, then the text in the box will disappear
  - If a user has an option selected, and moves to another phrase, but doesn't select it, and leaves the combo box, then the previous selection will remain
  - The Error State will meet accessibility requirements when placed against a dark theme background - see comp

### Storybook

* 📁 Forms
   * 📁 Form elements
      * 📁 Form controls
         * 📁 Combo box
            * 📁 Default
               * 📁 **Default state**
               * 📁 **Disabled state**
               * 📁 **No JS**
            * 📁 Test cases
               * 📁 Cgov
                  * **Cgov input**
                  * **Cgov input error**

### Notes

- [x] ACTION ITEM: @arcepaul will review the issues related to the USWDS combo box to find what fixes should be made
- [x] ACTION ITEM: @karlikpj to determine what the error state is for a combo box with the outline from NCIDS/USWDS
- [x] ACTION ITEM: @blilianyu @monika-jaeger @dcvijanovic1k will update the comps to match the functional requirements listed above, include changing the third comp image to reflect the error state if a user presses 'Go' without an entry in the combo box.

### DoR DS Tasks

- [x] Update design of nci-combo-box for all visual and functional requirements identified above. Put the solution into its own class space on Figma
- [x] Refine user story statement
- [x]  Create Scenarios
- [ ]  Internal review of any changes to the Design solution for nci-combo-box on Figma
  - [x] Design Approval
  - [x] Development Approval
  - [x] IA Approval
- [ ]  Internal review of DS AC's
  - [x] Design Approval
  - [x] Development Approval
  - [x] IA Approval

### DoD

- [ ] Develop DS class from design Figma comp
  - [ ]  Update Documentation Site per Scenarios document
    - [ ]  **To be completed by #1409 :** Draft the guidance and all sections necessary for addition to the NCIDS for All Documentation Site
    - [ ] Create page for the Combo Box under the Components section in Doc Site
    - [ ] Update the [Components landing page](https://designsystem-dev.cancer.gov/develop/components/) so that the link to the Combo Box is on this page
  - [ ]  Update Storybook Site per Scenarios document
  - [ ]  Update Example Site per Scenarios document
- [ ]  Internal review of the DS solution
- [ ]  Demo DS solution to client

### Resources:
* [Cancer Types Comp](https://www.figma.com/file/ZIgPLcQOlHYrbUxeWYaanZ/Cancer-Types?type=design&node-id=0-1&mode=design&t=kO43db8PZZekaRcW-0)
* [Combo Box Comp](https://www.figma.com/file/3LA0fAGyGdzpBeNDCoulFf/NCIDS-Design-Kit---Master?type=design&node-id=5869-11262&mode=design)
* [Combo Box Error State Comp](https://www.figma.com/file/3LA0fAGyGdzpBeNDCoulFf/NCIDS-Design-Kit---Master?type=design&node-id=5869-11262&mode=design)
* [Scenarios](https://nih.sharepoint.com/:w:/r/sites/NCI-OCPLNewCgovUXDesignSystem/Shared%20Documents/General/Details%20-%20Release%202/R2.4/Development%20and%20Testing/NCI%20Combo%20Box%20Scenarios.docx?d=w34abae13b5d34917bff344eddfeafd44&csf=1&web=1&e=ZqKMvb)
* [Doc Site Content](https://nih.sharepoint.com/:w:/r/sites/NCI-OCPLNewCgovUXDesignSystem/_layouts/15/Doc.aspx?sourcedoc=%7B364941EA-078F-465F-8029-B53C74D286D5%7D&file=Combo%20Box.docx&wdLOR=cFB8AD81E-6B6F-E949-8089-D9CA2476D443&action=default&mobileredirect=true)

### Acceptance Criteria

```gherkin
Scenario: A user views a page with an nci-combo-box component
  Given a page editor has added an nci-combo-box component to the page
    And the nci-combo-box component has been implemented in the Design System
    And the nci-combo-box has a list of items to pull from
  When the user views the page with an nci-combo-box component
  Then the nci-combo-box will match the comp (see link in the Resources section)
  When the user clicks in the combo box
  Then the combo box will be outlined to indicate it has been clicked
    And the cursor will appear in the combo box
  When the user clicks the down arrow in the combo box
  Then all the options in the list of items will appear in alphabetical order
  When the user types a string of characters in the combo box
  Then the dropdown will display any options that contain that string (ignoring capitalization) in alphabetical order
  When a user types in a partial string into the combo box
    And presses enter
  Then no selection will be made
  When a user types a string into the combo box that is not contained within any of the options of the dropdown
  Then the only item in the dropdown will be "No results found"
  When a user types a fully matching string into the combo box
    And presses enter
  Then the matching string will populate the combo box
  When a user clicks on an option from the dropdown list
  Then the selection will populate the combo box
  When there is a selection populating the combo box
  Then an 'X' will appear at the right side of the combo box
  When there is a selection populating the combo box
    And the user clicks the 'X'
  Then the combo box will be cleared
  When there is a selection populating the combo box
    And the user backspaces/deletes/highlights and writes over the text in the combo box
    And presses enter
  Then the previous selection populating the combo box will remain populating the combo box
  When the user tabs to the combo box
  Then the combo box will be outlined to indicate it has been clicked
    And the cursor will appear in the combo box
  When the user presses the down arrow
  Then the dropdown will appear
    And the first option in the dropdown will be outlined
  When the user presses the down arrow N times
  Then the Nth option in the dropdown will be outlined
  When the user presses the up arrow
    And the user is on an item in the dropdown
    And is not in the text area of the combo box
  Then they will move up one item in the list
  When an item is outlined via keyboard navigation
    And the user presses enter
  Then the item that was outlined will populate the combo box as a selection
  When the dropdown list is visible to the user
    And the user presses the Esc key
  Then the dropdown list will collapse
  When the user has the combo box text area selected
    And there is text populating the combo box
    And the cursor is not at the far left (beginning of the populated word/phrase)
    And the user presses the left arrow key
  Then the cursor will move to the left by one character
  When the user has the combo box text area selected
    And there is text populating the combo box
    And the cursor is not at the far right (end of the populated word/phrase)
    And the user presses the right arrow key
  Then the cursor will move to the right by one character
  When the user selects the combo box
    And there is no text populating the combo box
    And the user presses the tab key
  Then the selection will move to the next object within the structure of the page - not in the combo box
  When the user selects the combo box
    And there is text populating the combo box
    And the user presses the tab key
  Then the hover state will move to the 'X'
  When the hover state is over the 'X'
    And the user presses the enter key
  Then the text in the combo box will be cleared
  When the hover state is over the 'X'
    And the user presses the tab key
  Then the selection will move to the next object within the structure of the page - not in the combo box
```

Combo box - Disabled State
```gherkin
Scenario: A user views a page with an nci-combo-box component in the disabled state
  Given a page editor has added an nci-combo-box component to the page
    And the nci-combo-box component has been implemented in the Design System
    And the nci-combo-box has a list of items to pull from
    And the nci-combo-box is in the disabled state
  When the user views the page with an nci-combo-box component
  Then the combo box will be greyed out
    And the user will not be able to interact with the combo box
```
### Technical Notes
- [Editable combobox example](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/)
- W3C Recommended Keyboard interactions - [listbox example](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/)
- Open USWDS combo box issues: [https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+combo+box](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+combo+box)

### PR DESCRIPTION
**Pull request by olitharp-nci**
Closes #1397

---

# A11y checklist

Please see [comment](https://github.com/NCIOCPL/ncids/pull/1521#issuecomment-1960377025) for blank template.

## Keyboard Interaction

- [x] `Tab`: The combobox is in the page `Tab` sequence.
- [x] Note: The popup indicator icon or button (if present), the popup, and the popup descendants are excluded from the page `Tab` sequence.

### Combobox Keyboard Interaction

When focus is in the combobox:

- [x] `Down Arrow`: If the popup is available, moves [visual] focus into the popup:
  - [x] If the autocomplete behavior automatically selected a suggestion before `Down Arrow` was pressed, [visual] focus is placed on the suggestion following the automatically selected suggestion.
  - [x] Otherwise, places [visual] focus on the first focusable element in the popup.
- [x] `Up Arrow` (Optional): If the popup is available, places [visual] focus on the last focusable element in the popup.
- [x] `Escape`: Dismisses the popup if it is visible. Optionally, if the popup is hidden before `Escape` is pressed, clears the combobox.
- [x] `Enter`: If the combobox is editable and an autocomplete suggestion is selected in the popup, accepts the suggestion either by placing the input cursor at the end of the accepted value in the combobox or by performing a default action on the value. For example, in a messaging application, the default action may be to add the accepted value to a list of message recipients and then clear the combobox so the user can add another recipient.
- [x] Printable Characters:
  - [x] If the combobox is editable, type characters in the combobox. Note that some implementations may regard certain characters as invalid and prevent their input.
- [x] If the combobox is editable, it supports standard single line text editing keys appropriate for the device platform.

### Listbox Popup Keyboard Interaction

When focus is in a listbox popup:

- [x] `Enter`: Accepts the focused option in the listbox by closing the popup, placing the accepted value in the combobox, and if the combobox is editable, placing the input cursor at the end of the value.
- [x] `Escape`: Closes the popup and returns focus to the combobox. Optionally, if the combobox is editable, clears the contents of the combobox.
- [x] `Down Arrow`: Moves focus to and selects the next option. If focus is on the last option, either returns focus to the combobox or does nothing.
- [x] `Up Arrow`: Moves focus to and selects the previous option. If focus is on the first option, either returns focus to the combobox or does nothing.
- [x] `Right Arrow`: If the combobox is editable, returns focus to the combobox without closing the popup and moves the input cursor one character to the right. If the input cursor is on the right-most character, the cursor does not move.
- [x] `Left Arrow`: If the combobox is editable, returns focus to the combobox without closing the popup and moves the input cursor one character to the left. If the input cursor is on the left-most character, the cursor does not move.
- [x] Any printable character:
  - [x] If the combobox is editable, returns the focus to the combobox without closing the popup and types the character.
  - [x] Otherwise, moves focus to the next option with a name that starts with the characters typed.

### Combobox WAI-ARIA Roles, States, and Properties

- [x] The element that serves as an input and displays the combobox value has role [`combobox`](https://w3c.github.io/aria/#combobox).
- [x] The combobox element has [`aria-controls`](https://w3c.github.io/aria/#aria-controls) set to a value that refers to the element that serves as the popup. Note that `aria-controls` only needs to be set when the popup is visible. However, it is valid to reference an element that is not visible.
- [x] The popup is an element that has role [`listbox`](https://w3c.github.io/aria/#listbox), [`tree`](https://w3c.github.io/aria/#tree), [grid](https://w3c.github.io/aria/#grid), or [`dialog`](https://w3c.github.io/aria/#dialog).
- [x] When the combobox popup is not visible, the element with role `combobox` has [`aria-expanded`](https://w3c.github.io/aria/#aria-expanded) set to `false`. When the popup element is visible, `aria-expanded` is set to `true`. Note that elements with role `combobox` have a default value for `aria-expanded` of `false`.
- [x] When a combobox receives focus, DOM focus is placed on the combobox element.
- [x] When a descendant of a listbox, grid, or tree popup is focused, DOM focus remains on the combobox and the combobox has [`aria-activedescendant`](https://w3c.github.io/aria/#aria-activedescendant) set to a value that refers to the focused element within the popup.
- [x] For a combobox that controls a listbox, grid, or tree popup, when a suggested value is visually indicated as the currently selected value, the `option`, `gridcell`, `row`, or `treeitem` containing that value has [`aria-selected`](https://w3c.github.io/aria/#aria-selected) set to `true`.
- [x] If the combobox has a visible label and the combobox element is an HTML element that can be labelled using the HTML `label` element (e.g., the `input` element), it is labeled using the `label` element. Otherwise, if it has a visible label, the combobox element has [`aria-labelledby`](https://w3c.github.io/aria/#aria-labelledby) set to a value that refers to the labelling element. Otherwise, the combobox element has a label provided by [`aria-label`](https://w3c.github.io/aria/#aria-label).
- [x] The combobox element has [`aria-autocomplete`](https://w3c.github.io/aria/#aria-autocomplete) set to a value that corresponds to its autocomplete behavior:
  - [x] `list`: When the popup is triggered, it presents suggested values. If the combobox is editable, the values complete or logically correspond to the characters typed in the combobox.

### Listbox WAI-ARIA Roles, States, and Properties

- [x] An element that contains or owns all the listbox options has role [`listbox`](https://w3c.github.io/aria/#listbox).
- [x] Each option in the listbox has role [`option`](https://w3c.github.io/aria/#option) and is contained in or owned by either:
  - [x] The element with role `listbox`.
- [x] If the element with role `listbox` is not part of another widget, such as a combobox, then it has either a visible label referenced by [`aria-labelledby`](https://w3c.github.io/aria/#aria-labelledby) or a value specified for [`aria-label`](https://w3c.github.io/aria/#aria-label).
- [x] The selection state of each selectable option is indicated with either [`aria-selected`](https://w3c.github.io/aria/#aria-selected) or [`aria-checked`](https://w3c.github.io/aria/#aria-checked):
  - [x] If the selection state is indicated with `aria-selected`, then `aria-checked` is not specified for any options. Alternatively, if the selection state is indicated with `aria-checked`, then `aria-selected` is not specified for any options.
  - [x] If any options are selected, each selected option has either [`aria-selected`](https://w3c.github.io/aria/#aria-selected) or [`aria-checked`](https://w3c.github.io/aria/#aria-checked) set to `true`. No more than one option is selected at a time.
  - [x] All options that are selectable but not selected have either [`aria-selected`](https://w3c.github.io/aria/#aria-selected) or [`aria-checked`](https://w3c.github.io/aria/#aria-checked) set to `false`.
  - [x] Note that except in listboxes where selection follows focus, the selected state is distinct from focus. For more details, see [this description of differences between focus and selection](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#kbd_focus_vs_selection) and [Deciding When to Make Selection Automatically Follow Focus](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#kbd_selection_follows_focus).
- [x] If the complete set of available options is not present in the DOM due to dynamic loading as the user scrolls, their [`aria-setsize`](https://w3c.github.io/aria/#aria-setsize) and [`aria-posinset`](https://w3c.github.io/aria/#aria-posinset) attributes are set appropriately.

### CODE REVIEW

**Comment by olitharp-nci** [comment]
Changes from USWDS combo box

[Keyboard events](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-both/#kbd_label)

1. Textbox: Arrow down

   > * If the listbox is displayed and a suggestion is selected, moves visual focus to the next suggested value.
   > * If the textbox is empty and the listbox is not displayed, opens the listbox and moves visual focus to the first option.
   > * In both cases DOM focus remains on the textbox.
   
   Focus must remain in the textbox so that the user can still type. USWDS combo box moves all focus to the next suggested value.

2. Textbox: Alt  + Arrow Down

   > Opens the listbox without moving focus or changing selection.
   
   Focus must remain the the textbox so that the user can style type. Holding alt should allow the textbox to scroll without changing visual focus or changing the selected option. USWDS sets focus to the first option in the list if no section is made. If a selection is made, the keyboard event works as expected.

3. Textbox: Arrow Up

   > * If the listbox is displayed and a suggestion is selected, moves visual focus to the last suggested value.
   > * If the textbox is empty, first opens the listbox if it is not already displayed and then moves visual focus to the last option.
   > * In both cases DOM focus remains on the textbox.
   
   Similar to Arrow Down, Arrow Up has the same expectations, except it visually moves the focus to the last option. USWDS does not support Arrow Up.

**Comment by olitharp-nci** [comment]
# A11y checklist

* A "[combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox)" is defined as a "input widget" with an associated popup. 
* A "[listbox](https://www.w3.org/WAI/ARIA/apg/patterns/listbox)" is defined as a list of options that a user may select one more multiple.

## Keyboard Interaction

- [ ] `Tab`: The combobox is in the page `Tab` sequence.
- [ ] Note: The popup indicator icon or button (if present), the popup, and the popup descendants are excluded from the page `Tab` sequence.

### Combobox Keyboard Interaction

When focus is in the combobox:

- [ ] `Down Arrow`: If the popup is available, moves focus into the popup:
  - [ ] If the autocomplete behavior automatically selected a suggestion before `Down Arrow` was pressed, focus is placed on the suggestion following the automatically selected suggestion.
  - [ ] Otherwise, places focus on the first focusable element in the popup.
- [ ] `Up Arrow` (Optional): If the popup is available, places focus on the last focusable element in the popup.
- [ ] `Escape`: Dismisses the popup if it is visible. Optionally, if the popup is hidden before `Escape` is pressed, clears the combobox.
- [ ] `Enter`: If the combobox is editable and an autocomplete suggestion is selected in the popup, accepts the suggestion either by placing the input cursor at the end of the accepted value in the combobox or by performing a default action on the value. For example, in a messaging application, the default action may be to add the accepted value to a list of message recipients and then clear the combobox so the user can add another recipient.
- [ ] Printable Characters:
  - [ ] If the combobox is editable, type characters in the combobox. Note that some implementations may regard certain characters as invalid and prevent their input.
  - [ ] If the combobox is not editable, optionally moves focus to a value that starts with the typed characters.
- [ ] If the combobox is editable, it supports standard single line text editing keys appropriate for the device platform (see note below).
- [ ] `Alt + Down Arrow` (Optional): If the popup is available but not displayed, displays the popup without moving focus.
- [ ] `Alt + Up Arrow` (Optional): If the popup is displayed:
  - [ ] If the popup contains focus, returns focus to the combobox.
  - [ ] Closes the popup.

### Listbox Popup Keyboard Interaction

When focus is in a listbox popup:

- [ ] `Enter`: Accepts the focused option in the listbox by closing the popup, placing the accepted value in the combobox, and if the combobox is editable, placing the input cursor at the end of the value.
- [ ] `Escape`: Closes the popup and returns focus to the combobox. Optionally, if the combobox is editable, clears the contents of the combobox.
- [ ] `Down Arrow`: Moves focus to and selects the next option. If focus is on the last option, either returns focus to the combobox or does nothing.
- [ ] `Up Arrow`: Moves focus to and selects the previous option. If focus is on the first option, either returns focus to the combobox or does nothing.
- [ ] `Right Arrow`: If the combobox is editable, returns focus to the combobox without closing the popup and moves the input cursor one character to the right. If the input cursor is on the right-most character, the cursor does not move.
- [ ] `Left Arrow`: If the combobox is editable, returns focus to the combobox without closing the popup and moves the input cursor one character to the left. If the input cursor is on the left-most character, the cursor does not move.
- [ ] `Home` (Optional): Either moves focus to and selects the first option or, if the combobox is editable, returns focus to the combobox and places the cursor on the first character.
- [ ] `End` (Optional): Either moves focus to the last option or, if the combobox is editable, returns focus to the combobox and places the cursor after the last character.
- [ ] Any printable character:
  - [ ] If the combobox is editable, returns the focus to the combobox without closing the popup and types the character.
  - [ ] Otherwise, moves focus to the next option with a name that starts with the characters typed.
- [ ] `Backspace` (Optional): If the combobox is editable, returns focus to the combobox and deletes the character prior to the cursor.
- [ ] `Delete` (Optional): If the combobox is editable, returns focus to the combobox, removes the selected state if a suggestion was selected, and removes the inline autocomplete string if present.

## WAI-ARIA Roles, States, and Properties

### Combobox WAI-ARIA Roles, States, and Properties

- [ ] The element that serves as an input and displays the combobox value has role [`combobox`](https://w3c.github.io/aria/#combobox).
- [ ] The combobox element has [`aria-controls`](https://w3c.github.io/aria/#aria-controls) set to a value that refers to the element that serves as the popup. Note that `aria-controls` only needs to be set when the popup is visible. However, it is valid to reference an element that is not visible.
- [ ] The popup is an element that has role [`listbox`](https://w3c.github.io/aria/#listbox), [`tree`](https://w3c.github.io/aria/#tree), [grid](https://w3c.github.io/aria/#grid), or [`dialog`](https://w3c.github.io/aria/#dialog).
- [ ] If the popup has a role other than `listbox`, the element with role `combobox` has [`aria-haspopup`](https://w3c.github.io/aria/#aria-haspopup) set to a value that corresponds to the popup type. That is, `aria-haspopup` is set to `grid`, `tree`, or `dialog`. Note that elements with role `combobox` have an implicit `aria-haspopup` value of `listbox`.
- [ ] When the combobox popup is not visible, the element with role `combobox` has [`aria-expanded`](https://w3c.github.io/aria/#aria-expanded) set to `false`. When the popup element is visible, `aria-expanded` is set to `true`. Note that elements with role `combobox` have a default value for `aria-expanded` of `false`.
- [ ] When a combobox receives focus, DOM focus is placed on the combobox element.
- [ ] When a descendant of a listbox, grid, or tree popup is focused, DOM focus remains on the combobox and the combobox has [`aria-activedescendant`](https://w3c.github.io/aria/#aria-activedescendant) set to a value that refers to the focused element within the popup.
- [ ] For a combobox that controls a listbox, grid, or tree popup, when a suggested value is visually indicated as the currently selected value, the `option`, `gridcell`, `row`, or `treeitem` containing that value has [`aria-selected`](https://w3c.github.io/aria/#aria-selected) set to `true`.
- [ ] If the combobox has a visible label and the combobox element is an HTML element that can be labelled using the HTML `label` element (e.g., the `input` element), it is labeled using the `label` element. Otherwise, if it has a visible label, the combobox element has [`aria-labelledby`](https://w3c.github.io/aria/#aria-labelledby) set to a value that refers to the labelling element. Otherwise, the combobox element has a label provided by [`aria-label`](https://w3c.github.io/aria/#aria-label).
- [ ] The combobox element has [`aria-autocomplete`](https://w3c.github.io/aria/#aria-autocomplete) set to a value that corresponds to its autocomplete behavior:
  - [ ] `none`: When the popup is displayed, the suggested values it contains are the same regardless of the characters typed in the combobox.
  - [ ] `list`: When the popup is triggered, it presents suggested values. If the combobox is editable, the values complete or logically correspond to the characters typed in the combobox.
  - [ ] `both`: When the popup is triggered, it presents suggested values that complete or logically correspond to the characters typed in the combobox. In addition, the portion of the selected suggestion that has not been typed by the user, known as the completion string, appears inline after the input cursor in the combobox. The inline completion string is visually highlighted and has a selected state.

### Listbox WAI-ARIA Roles, States, and Properties

- [ ] An element that contains or owns all the listbox options has role [`listbox`](https://w3c.github.io/aria/#listbox).
- [ ] Each option in the listbox has role [`option`](https://w3c.github.io/aria/#option) and is contained in or owned by either:
  - [ ] The element with role `listbox`.
  - [ ] An element with role [`group`](https://w3c.github.io/aria/#group) that is contained in or owned by the element with role `listbox`.
- [ ] Options contained in a group are referred to as grouped options and form what is called an option group. If a listbox contains grouped options, then:
  - [ ] All option groups contain at least one option.
  - [ ] Each option group has an accessible name provided via [aria-label](https://w3c.github.io/aria/#aria-label) or [`aria-labelledby`](https://w3c.github.io/aria/#aria-labelledby).
- [ ] If the element with role `listbox` is not part of another widget, such as a combobox, then it has either a visible label referenced by [`aria-labelledby`](https://w3c.github.io/aria/#aria-labelledby) or a value specified for [`aria-label`](https://w3c.github.io/aria/#aria-label).
- [ ] If the listbox supports selection of more than one option, the element with role `listbox` has [aria-multiselectable](https://w3c.github.io/aria/#aria-multiselectable) set to true. Otherwise, `aria-multiselectable` is either set to `false` or the default value of `false` is implied.
- [ ] The selection state of each selectable option is indicated with either [`aria-selected`](https://w3c.github.io/aria/#aria-selected) or [`aria-checked`](https://w3c.github.io/aria/#aria-checked):
  - [ ] If the selection state is indicated with `aria-selected`, then `aria-checked` is not specified for any options. Alternatively, if the selection state is indicated with `aria-checked`, then `aria-selected` is not specified for any options. See notes below regarding considerations for which property to use and for details of the unusual conditions that might allow for both properties in the same listbox.
  - [ ] If any options are selected, each selected option has either [`aria-selected`](https://w3c.github.io/aria/#aria-selected) or [`aria-checked`](https://w3c.github.io/aria/#aria-checked) set to `true`. No more than one option is selected at a time if the element with role `listbox` does not have [`aria-multiselectable`](https://w3c.github.io/aria/#aria-multiselectable) set to true.
  - [ ] All options that are selectable but not selected have either [`aria-selected`](https://w3c.github.io/aria/#aria-selected) or [`aria-checked`](https://w3c.github.io/aria/#aria-checked) set to `false`.
  - [ ] Note that except in listboxes where selection follows focus, the selected state is distinct from focus. For more details, see [this description of differences between focus and selection](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#kbd_focus_vs_selection) and [Deciding When to Make Selection Automatically Follow Focus](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#kbd_selection_follows_focus).
- [ ] If the complete set of available options is not present in the DOM due to dynamic loading as the user scrolls, their [`aria-setsize`](https://w3c.github.io/aria/#aria-setsize) and [`aria-posinset`](https://w3c.github.io/aria/#aria-posinset) attributes are set appropriately.
- [ ] If options are arranged horizontally, the element with role listbox has [`aria-orientation`](https://w3c.github.io/aria/#aria-orientation) set to `horizontal`. The default value of `aria-orientation` for `listbox` is `vertical`.

**Comment by blilianyu** [comment]
@olitharp-nci one minor thing, the dropdown is wider than the input on all breakpoints
Screenshot 2024-03-27 at 9 45 43 AM

**Comment by adrianacastaneda** [comment]
If possible, let's remove this line from the Character Limits section: "Character limits are based on best practices and when a line of text will be forced to wrap onto a second line of text on desktop."  I missed this initially. We usually have this line there but it's not relevant to this component. Thank you!

**Comment by dcvijanovic1k** [comment]
Red line on error state showing up inconsistently or not at all across all browsers (FF, Chrome, Safari). Images showing error in Chrome and Safari attached.
Screenshot 2024-03-27 at 10 21 32 AM
Screenshot 2024-03-27 at 10 21 11 AM

**Comment by andyvanavery31** [comment]
@olitharp-nci please remove the "Select a Fruit" option from the combo box on the documentation site.

For Dragutin's previous comment, I am seeing the red line appear on the left side of the combo box, when it is viewed at a breakpoint less than desktop (<1024px) and is not in a container. For example, it occurs on the Error State in the Default Folder, but does not appear at any breakpoint in the Cgov Input Error story within the Test Cases folder.

Could you please also investigate the following things related to No Results found?

- When the string entered matches none of the options, please change the text that appears in the dropdown from "No Results" to "No Results Found"
- When the user tabs into the combobox, then types in a random string(the No Results Found option appears), and they press enter, the first item in the dropdown list gets selected
- When the user selects the combobox and the dropdown list appears, then the user hovers over an item on the dropdown list with their cursor, and then types in a random string (the No Results Found option appears), and they press enter, the item that was hovered before the dropdown collapsed to "No Results Found" over gets selected

**Comment by olitharp-nci** [comment]
> If possible, let's remove this line from the Character Limits section: "Character limits are based on best practices and when a line of text will be forced to wrap onto a second line of text on desktop." I missed this initially. We usually have this line there but it's not relevant to this component. Thank you!

@adrianacastaneda That has been removed.

> @olitharp-nci one minor thing, the dropdown is wider than the input on all breakpoints Screenshot 2024-03-27 at 9 45 43 AM

@blilianyu thanks for finding this! This should now be fixed, I have added a regression test for this as well. 

> Red line on error state showing up inconsistently or not at all across all browsers (FF, Chrome, Safari). Images showing error in Chrome and Safari attached. Screenshot 2024-03-27 at 7 18 39 AM 

@dcvijanovic1k and @andyvanavery31  That tricksy little left border... You can see the same issue on with USWDS [here](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/develop/?path=/story/components-form-inputs-file-input--error) and [here](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/develop/?path=/story/components-form-inputs-text-input--state-showcase). I believe uswds issues 2476 and 5839 are related, but not completely solve this in a timely manner. We _may_ be able to put in a patch to fix this in the meantime. I'll let y'all and @bennettcc decide on the best course of action forward on this.

@andyvanavery31 "No results found" has been updated, but I'm still working through the other 3 issues. 

Thanks all. Will update my progress report in the next push.

File: packages/ncids-twig/components/usa-combo-box/usa-combo-box.twig [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<div class="usa-form-group {% if error %}usa-form-group--error{% endif %} {{ form_group_class }}">
+	<label class="usa-label {{ label_class }}" for="{{ name }}">{{ label }}</label>
+	{% if error %}
+		<span class="usa-error-message {{ error_message_class }}">Please enter a cancer type</span>
```
Comment (sarinapadilla) [inline packages/ncids-twig/components/usa-combo-box/usa-combo-box.twig:4]: Change this to "Please enter a fruit" please!

**Comment by monika-jaeger** [comment]
The rest looks good from design!

**Comment by bennettcc** [comment]
Federal team has approved the Combo box - tested by Alex and Jorge on 3/29

**Review by sarinapadilla — APPROVED** [review]
"Please select a fruit" - looks good!!

**Review by arcepaul — APPROVED** [review]

File: testing/ncids-css-testing/stories/uswds-native/usa-combo-box/src/index.scss [hunk 2]
```diff
@@ -1,5 +1,3 @@
 @use "styles/ncids";
 @forward "usa-label";
```
Comment (bryanpizzillo) [inline testing/ncids-css-testing/stories/uswds-native/usa-combo-box/src/index.scss:2]: This is just a question: why did these get removed?

File: testing/ncids-js-testing/config/webpack.dev.js [hunk 3]
```diff
@@ -15,7 +15,7 @@ module.exports = merge(common, {
 	devServer: {
 		historyApiFallback: true,
 		static: paths.public,
-		open: true,
+		open: false,
```
Comment (bryanpizzillo) [inline testing/ncids-js-testing/config/webpack.dev.js:18]: Did you mean to commit this? (Frankly I lothe the auto open, so I am all in favor of this)

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
There are a couple of things. The event stuff we can negotiate.

File: testing/ncids-css-testing/stories/components/usa-combo-box/src/usa-combo-box.scenarios.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
+module.exports = [
+	// default
+	{
```
[the remark this answers is not in the document]
Comment (olitharp-nci) [reply testing/ncids-css-testing/stories/components/usa-combo-box/src/usa-combo-box.scenarios.js:3]: adding mobile-lg, tablet, tablet-lg, and widescreen to cgov and cgov error

File: testing/ncids-css-testing/stories/uswds-native/usa-combo-box/src/index.scss (hunk 2 above)
Comment (olitharp-nci) [reply testing/ncids-css-testing/stories/uswds-native/usa-combo-box/src/index.scss:2]: https://github.com/uswds/uswds/blob/e901bc69bd09adb07a7fefd0ff53751adf811bff/packages/usa-combo-box/_index.scss#L2

redundant. imported in the package.

File: testing/ncids-js-testing/config/webpack.dev.js (hunk 3 above)
Comment (olitharp-nci) [reply testing/ncids-js-testing/config/webpack.dev.js:18]: no but can i keep it? i'll feed it and wash it and pick up after it and pay for all the vet bills

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
Can you only set autocomplete=off if the select in the dom has autocomplete=off

**EDIT**: We talked about this and it is not a good idea. See Olivia's known issues below about the back button for what we actually found. (e.g. USWDS does this and Safari is annoying)

**Comment by olitharp-nci** [comment]
# Known issues

- [ ] Combo box option values must be unique, these should be optional.

   Due to setting `HTMLSelectElement` through javascript, `HTMLSelectElement.value = '...'`,  the option values must be unique. Switching to set `HTMLOptionElement[index].selected` may be a valid workaround.

- [ ] Combo box behavior using the back button on browsers may differ.

   After form submission, click the back button on the browser causes the form element to either be autocompleted with the previous value. This behavior differs due to the `autocomplete="off"` attribute on the `HTMLSelectElement`:
   
     * Chrome values are removed.
     * Firefox values are retained. 
     * Safari values are removed, however, the `.combo-box--pristine` class remains, which causes the clear button to be shown.
      
   This behavior can be repeated using the USWDS combo box inside of a form.

**Review by bryanpizzillo — APPROVED** [review]

**Comment by bennettcc** [comment]
The issues reported above have been created as standalone backlog items for future fixing

**Pull request merged**